### PR TITLE
Feature/engine data proxy

### PR DESCRIPTION
**Pull request by sergiimk**
In this PR:
- Isolated Ingest/Transform services from FS paths
  - How data gets into the engine is now decided on the engine level
  - There will be several ways to pass data in and out of the engine in future (mounting files, direct remote storage links, via shared memory)
  - This is the layer I will work on next, to make ingest/transform work for S3 datasets (initially without touching the engines)
- These changes allowed to better formalize the commit procedure
- AND to finally move `WorkspaceLayout` to app layer
- Also we're no longer mounting entire workspace into engine containers:
  - most things are mounted on per-file basis (except ingest where we still need to read all data directory)
  - all inputs are mounted as Read-Only for safety

### CODE REVIEW

File: src/domain/core/src/entities/dataset.rs [hunk 1]
```diff
@@ -148,3 +165,59 @@ pub enum CommitError {
         InternalError,
     ),
 }
+
+/////////////////////////////////////////////////////////////////////////////////////////
+// Commit helpers
+/////////////////////////////////////////////////////////////////////////////////////////
+
+/// Replicates [AddData] event prior to hasing of data and checkpoint
```
Comment (zaychenko-sergei) [inline src/domain/core/src/entities/dataset.rs:173]: prior to has**h**ing

File: src/domain/core/src/entities/dataset.rs [hunk 2]
```diff
@@ -148,3 +165,59 @@ pub enum CommitError {
...
+    path: Option<PathBuf>,
+}
+
+impl OwnedFile {
+    pub fn new(path: impl Into<PathBuf>) -> Self {
+        let path = path.into();
+        assert!(path.exists());
+        Self { path: Some(path) }
+    }
+
```
Comment (zaychenko-sergei) [inline src/domain/core/src/entities/dataset.rs:207]: I guess this should be isolated in `domain/core/src/utils` rather then staying in the `dataset.rs`

File: src/infra/core/src/engine/engine_io_strategy.rs [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+    async fn materialize_object(
+        &self,
+        repo: &dyn ObjectRepository,
+        hash: &Multihash,
+        container_in_dir: &Path,
+        volumes: &mut Vec<VolumeSpec>,
+    ) -> Result<PathBuf, InternalError> {
+        let url = repo.get_internal_url(hash).await;
+        let host_path = kamu_data_utils::data::local_url::into_local_path(url).int_err()?;
```
Comment (zaychenko-sergei) [inline src/infra/core/src/engine/engine_io_strategy.rs:39]: Should we use `CachedObject` here maybe?

File: src/domain/core/src/entities/dataset.rs (hunk 1 above)
Comment (sergiimk) [reply src/domain/core/src/entities/dataset.rs:173]: thanks, fixed

File: src/domain/core/src/entities/dataset.rs (hunk 2 above)
Comment (sergiimk) [reply src/domain/core/src/entities/dataset.rs:207]: done!

File: src/infra/core/src/engine/engine_io_strategy.rs (hunk 3 above)
Comment (sergiimk) [reply src/infra/core/src/engine/engine_io_strategy.rs:39]: Yep, this will likely be a part of my next PR

**Pull request merged**